### PR TITLE
refactor(settings): adjust assertion caching

### LINKED ISSUE #4949: Refreshing device manager after several hours is invalid
STEPS TO REPRODUCE
1. Log in to device manager
2. Leave tab open
3. Go to work
4. Perhaps visit device manager there
5. Return home to still-open device manager tab and press Refresh

EXPECTED RESULTS
* Devices refresh

ACTUAL RESULTS
* Invalid assertion error
screenshot 2017-04-12 20 23 55

### PR DESCRIPTION
**Pull request by vladikoff**
Fixes #4949

@mozilla/fxa-devs r?

### CODE REVIEW

**Comment by rfk** [comment]
Nice :-)

I'm +1 to removing this, but is there anything specific we should measure to make sure it doesn't regress performance unacceptably?

**Comment by vladikoff** [comment]
> is there anything specific we should measure to make sure it doesn't regress performance unacceptably?

@rfk I think we would see a spike in OAuth Server logs and maybe go from there?

**Comment by shane-tomlinson** [comment]
> @rfk I think we would see a spike in OAuth Server logs and maybe go from there?

Does removing the cache negatively affect /settings performance since two assertions will be generated, one for each OAuth token? Looking in the console, I see two certificates are signed instead of one:

#### master

screen shot 2017-09-04 at 10 36 41

#### with patch

screen shot 2017-09-04 at 10 37 01

I can't tell fully, but it looks like the requests happen in parallel so maybe this isn't a big deal. Instead of disabling the cache, could we jettison entries after a short, known valid time - maybe 10 minutes?

I'm concerned about further degradation on the settings page if we can since it's already pretty bad.

ref https://github.com/mozilla/fxa-content-server/issues/5205

**Comment by vladikoff** [comment]
@shane-tomlinson updated to cache assertions for 1 hour

File: bower.json [hunk 1]
```diff
@@ -29,7 +29,7 @@
     "requirejs": "2.2.0",
     "requirejs-mustache": "https://github.com/jfparadis/requirejs-mustache.git#5fb8c0a3e8560526e8f9617a689e2924a8532d0a",
     "requirejs-text": "2.0.10",
-    "sinon": "1.17.5",
+    "sinon": "1.17.7",
```
Comment (vladikoff) [inline bower.json]: side note we should look into updating to sinon 3 at some point. Current version is having trouble with `fakeTimers` sadly.

File: app/tests/spec/models/account.js [hunk 2]
```diff
@@ -846,6 +846,19 @@ define(function (require, exports, module) {
           });
       });
 
+      it('generates a new assertion if assertion is expired', function () {
+        return account.createOAuthToken('scope1')
+          .then(function () {
+            sinon.stub(account, '_isAssertionValid', function () {
```
Comment (shane-tomlinson) [inline app/tests/spec/models/account.js]: You'll need to update this to the new stub type.

File: bower.json (hunk 1 above)
Comment (shane-tomlinson) [reply bower.json]: > side note we should look into updating to sinon 3 at some point

Your sinon 3 PR is merged!

File: app/scripts/models/account.js [hunk 3]
```diff
@@ -196,21 +196,30 @@ define(function (require, exports, module) {
...
+      assertionPromise.__createdAt = Date.now();
 
       this._assertionPromises[sessionToken] = assertionPromise;
 
       return assertionPromise;
     },
 
+    _isAssertionValid (assertionPromise) {
+      // assertions live for about 6 hours.
+      // reuse the same assertion if used in the past hour
```
Comment (shane-tomlinson) [inline app/scripts/models/account.js]: "used in the past hour" or "created in the past hour"?

Comment (vladikoff) [reply app/scripts/models/account.js]: true!

**Pull request merged**